Projects/Project_2/VirtualMachine/Library.fs

```diff
@@ -0,0 +1,163 @@
+  let sp = ref -1
+  let pc = ref 0 
+  let execStep i = 
+    executedInst <- executedInst + 1
+    match i with
+    | i when i=CODECSTI   -> s.[!sp+1] <- p.[!pc+1]; sp := !sp+1; pc := !pc+2
+    | i when i=CODEADD    -> s.[!sp-1] <- s.[!sp-1] + s.[!sp]; sp := !sp-1; pc := !pc+1
+    | i when i=CODESUB    -> s.[!sp-1] <- s.[!sp-1] - s.[!sp]; sp := !sp-1; pc := !pc+1
+    | i when i=CODEMUL    -> s.[!sp-1] <- s.[!sp-1] * s.[!sp]; sp := !sp-1; pc := !pc+1
+    | i when i=CODEDIV    -> s.[!sp-1] <- s.[!sp-1] / s.[!sp]; sp := !sp-1; pc := !pc+1
+    | i when i=CODEMOD    -> s.[!sp-1] <- s.[!sp-1] % s.[!sp]; sp := !sp-1; pc := !pc+1
+    | i when i=CODEEQ     -> s.[!sp-1] <- (if s.[!sp-1] = s.[!sp] then 1 else 0); sp := !sp-1; pc := !pc+1
+    | i when i=CODELT     -> s.[!sp-1] <- (if s.[!sp-1] < s.[!sp] then 1 else 0); sp := !sp-1; pc := !pc+1
+    | i when i=CODENOT    -> (s.[!sp] <- if s.[!sp] = 0 then 1 else 0); pc := !pc+1
+    | i when i=CODEDUP    -> s.[!sp+1] <- s.[!sp]; sp := !sp+1; pc := !pc+1
+    | i when i=CODESWAP   -> let tmp = s.[!sp]  
+                             s.[!sp] <- s.[!sp-1];  s.[!sp-1] <- tmp; pc := !pc+1 
+    | i when i=CODELDI    -> s.[!sp] <- s.[s.[!sp]]; pc := !pc+1                                  // load indirect        
+    | i when i=CODESTI    -> s.[s.[!sp-1]] <- s.[!sp]; s.[!sp-1] <- s.[!sp]; sp := !sp-1; pc := !pc+1 // store indirect, keep value on top
+    | i when i=CODEGETBP  -> s.[!sp+1] <- !bp; sp := !sp+1; pc := !pc+1
+    | i when i=CODEGETSP  -> s.[!sp+1] <- !sp; sp := !sp+1; pc := !pc+1
+    | i when i=CODEINCSP  -> sp := !sp+p.[!pc+1]; pc := !pc+2
+    | i when i=CODEGOTO   -> pc := p.[!pc+1]
+    | i when i=CODEIFZERO -> (pc := if s.[!sp] = 0 then p.[!pc+1] else !pc+2); sp := !sp-1
+    | i when i=CODEIFNZRO -> (pc := if s.[!sp] <> 0 then p.[!pc+1] else !pc+2); sp := !sp-1     
+    | i when i=CODECALL   -> 
+        let argc = p.[!pc+1]
+        let k = ref 0
+        while !k<argc do                             // Make room for return address
+          s.[!sp - !k+2] <- s.[!sp - !k]; k := !k+1  // and old base pointer
+        s.[!sp-argc+1] <- !pc+3; sp := !sp+1 
+        s.[!sp-argc+1] <- !bp; sp := !sp+1  
+        bp := !sp+1-argc
+        pc := p.[!pc+2] 
+    | i when i=CODETCALL  -> 
+        let argc = p.[!pc+1]                // Number of new arguments
+        let pop  = p.[!pc+2]               // Number of variables to discard
+        let k = ref (argc-1)
+        while !k>=0 do                    // Discard variables
+          s.[!sp - !k - pop] <- s.[!sp - !k]; k := !k-1 
+        sp := !sp - pop; pc := p.[!pc+3] 
+    | i when i=CODERET    -> 
+        let res = s.[!sp] 
+        sp := !sp-p.[!pc+1]; bp := s.[!sp-1]; pc := s.[!sp-2];  sp := !sp-2 
+        s.[!sp] <- res            
+    | i when i=CODEPRINTI -> System.Console.WriteLine(string s.[!sp] + " ") ; pc := !pc+1  
+    | i when i=CODEPRINTC -> System.Console.WriteLine((char)(s.[!sp])); pc := !pc+1 
```

Comment:
Not benchmarking, but comment: could we replace System.Console.WriteLine with System.Console.Write?